### PR TITLE
sql: fix prepare insert to specify return columns

### LINKED ISSUE #4604: sql: RETURNING causes pgwire panic
Adding a test like this in TestPGPreparedExec
```
466         {
467             "INSERT INTO d.T VALUES ($1) RETURNING 1",
468             []preparedExecTest{
469                 base.Params(1).RowsAffected(1),
470             },
471         },
```

Causes:

```
panic: runtime error: index out of range

goroutine 21058 [running]:
panic(0x160bfc0, 0xc82000e110)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).sendResponse(0xc8201841c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc8208f2a80, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:666 +0xd92
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc8201841c0, 0xc8201862b6, 0x27, 0xc820a8f240, 0x1, 0x1, 0x2578ba8, 0x0, 0x0, 0x0, ...)
```

@mjibson

### PR DESCRIPTION
**Pull request by madelynnblue**
Fixes #4604

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4635)

### CODE REVIEW

File: sql/insert.go [hunk 1]
```diff
@@ -247,7 +247,7 @@ func (p *planner) Insert(n *parser.Insert, autoCommit bool) (planNode, *roachpb.
 	}
 
 	if p.prepareOnly {
-		return nil, nil
+		return result, nil
```
Comment (petermattis) [inline sql/insert.go]: Add a comment explaining why this is needed.

**Comment by petermattis** [comment]
LGTM

File: sql/pgwire_test.go [hunk 2]
```diff
@@ -465,6 +465,12 @@ func TestPGPreparedExec(t *testing.T) {
 			},
 		},
 		{
+			"INSERT INTO d.T VALUES ($1) RETURNING 1",
```
Comment (tamird) [inline sql/pgwire_test.go]: i feel like this test case should be in the query block rather than the exec block

**Comment by RaduBerinde** [comment]
LGTM, thanks for fixing this so fast!

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/pgwire_test.go, line 467 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBOelvg76X5AZDvzmlN:-KBOelvg76X5AZDvzmlO:963014643) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c579c82ff53367b705c071f8660d4cf2f537aa6/sql/pgwire_test.go#L467)):*
Can you also add a couple of tests with `$1` inside RETURNING? (e.g. `RETURNING $i`, `RETURNING i + $i`). They should work (you might need to resync to master).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-:-KBOf1m_tI35x7UY-I35:1493708913)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/pgwire_test.go, line 467 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBOelvg76X5AZDvzmlN:-KBOf8vvf_88M-HV80T1:1171191321) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c579c82ff53367b705c071f8660d4cf2f537aa6/sql/pgwire_test.go#L467)):*
Sorry, meant `RETURNING $1`, `RETURNING i+$1`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by mjibson** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/insert.go, line 250 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBMRn_T2iBdhdOPXWyL-r1-250:-KBOxpBXdCx46wxgJNoU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c579c82ff53367b705c071f8660d4cf2f537aa6/sql/insert.go#L250)):*
Done.

---

*[sql/pgwire_test.go, line 467 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBOelvg76X5AZDvzmlN:-KBP1C0eLL36xDg_QVku:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c579c82ff53367b705c071f8660d4cf2f537aa6/sql/pgwire_test.go#L467)):*
Done.

---

*[sql/pgwire_test.go, line 468 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBMRn_T2iBdhdOPXWyM-r1-468:-KBP1CKd0b8t9M2vgfy8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c579c82ff53367b705c071f8660d4cf2f537aa6/sql/pgwire_test.go#L468)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/insert.go, line 251 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2UPYl4ypx3MY6A5h:-KBP2UPYl4ypx3MY6A5i:1841869700) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/insert.go#L251)):*
this is `rh.getResults()` now

---

*[sql/pgwire_test.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2M8OkAZXtT8khX9D:-KBP2M8OkAZXtT8khX9E:1748597774) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/pgwire_test.go#L325)):*
Are you sure this doesn't work on master? I suspect you are not synced to the latest changes

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by mjibson** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/insert.go, line 251 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2UPYl4ypx3MY6A5h:-KBP2rHfvj_ZPsPonhwU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/insert.go#L251)):*
Done.

---

*[sql/pgwire_test.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2M8OkAZXtT8khX9D:-KBP31xAU7bT4R0o8UYw:1645382783) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/pgwire_test.go#L325)):*
Yup, it doesn't work. It says:

INSERT INTO d.T VALUES ($1) RETURNING $1: [1]: unexpected error: pq: could not determine data type of parameter $1

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/pgwire_test.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2M8OkAZXtT8khX9D:-KBP3E_Hcfv1TvA_dXsx:600985875) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/pgwire_test.go#L325)):*
That's what I saw without PR #4609 but with those changes I thought I got as far as the panic..

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by knz** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/pgwire_test.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2M8OkAZXtT8khX9D:-KBPH0HgSmXpTsB4FFTN:620977379) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/pgwire_test.go#L325)):*
Hum wait I am not sure we want to support this specific syntax anyways. A placeholder in RETURNING is like trying to do SELECT $1 FROM ... which noone sane does. I'd suggest dropping this case entirely.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by mjibson** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/pgwire_test.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2M8OkAZXtT8khX9D:-KBPHR4np_vCI8V_6fG9:-287961168) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/pgwire_test.go#L325)):*
If it were just `INSERT INTO d.T Values(123) RETURNING $1` then I would agree. But in the test case here, we already know exactly the type of $1 since it's part of the VALUES statement. Thus there's no ambiguity and it could reasonably be supported.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635)*

**Comment by tamird** [comment]
LGTM

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/pgwire_test.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-KBP2M8OkAZXtT8khX9D:-KBPHhCjKwqJF48WXUIx:-351988714) ([raw file](https://github.com/cockroachdb/cockroach/blob/3160691ab41fd28cc5fb944ad6761f99276c932b/sql/pgwire_test.go#L325)):*
+1, there's no good reason to not support this case.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4635#-:-KBPHvIdaPKaTzB_gGC5:676834233)*

**Pull request merged**